binder/src/main/java/io/grpc/binder/internal/PendingAuthListener.java

```diff
@@ -0,0 +1,144 @@
+package io.grpc.binder.internal;
+
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import io.grpc.Metadata;
+import io.grpc.ServerCall;
+import io.grpc.ServerCallHandler;
+import io.grpc.Status;
+import io.grpc.internal.ObjectPool;
+
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicReference;
+
+import javax.annotation.Nullable;
+
+/**
+ * A {@link ServerCall.Listener} that can be returned by a {@link io.grpc.ServerInterceptor} to
+ * asynchronously advance the gRPC pending resolving a possibly asynchronous security policy check.
+ */
+final class PendingAuthListener<ReqT, RespT> extends ServerCall.Listener<ReqT> {
+
+  private final ConcurrentLinkedQueue<ListenerConsumer<ReqT>> pendingSteps =
+      new ConcurrentLinkedQueue<>();
+  private final AtomicReference<ServerCall.Listener<ReqT>> delegateRef =
+      new AtomicReference<>(null);
+
+  /**
+   * @param authStatusFuture a ListenableFuture holding the result status of the authorization
+   *                         policy from a {@link io.grpc.binder.SecurityPolicy} or a
+   *                         {@link io.grpc.binder.AsyncSecurityPolicy}. The call only progresses
+   *                         if {@link Status#isOk()} is true.
+   * @param executorPool     a pool that can provide at least one Executor under which the result
+   *                         of {@code authStatusFuture} can be handled, progressing the gRPC
+   *                         stages.
+   * @param call             the 'call' parameter from {@link io.grpc.ServerInterceptor}
+   * @param headers          the 'headers' parameter from {@link io.grpc.ServerInterceptor}
+   * @param next             the 'next' parameter from {@link io.grpc.ServerInterceptor}
+   */
+  PendingAuthListener(
+      ListenableFuture<Status> authStatusFuture,
+      ObjectPool<? extends Executor> executorPool,
+      ServerCall<ReqT, RespT> call,
+      Metadata headers,
+      ServerCallHandler<ReqT, RespT> next) {
+    Executor executor = executorPool.getObject();
+    Futures.addCallback(
+        authStatusFuture,
+        new FutureCallback<Status>() {
```

Comment:
Good catch. I think a `start()` method would make no difference since that guideline seems to be more about avoiding work in the constructor's call stack than just inside the constructor method.

I switched to a static factory method which addresses both issues.